docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,455 @@
+
+ovn-controller has this configuration, ovn-encap-tos:
+
+```text
+external_ids:ovn-encap-tos
+
+ovn-encap-tos indicates the value to be applied to OVN tunnel interface’s option:tos as specified in the Open_vSwitch database Interface table. Please refer to Open VSwitch Manual for details.
+
+   options : tos: optional string
+          Optional. The value of the ToS bits to be set on the encapsulat‐
+          ing packet. ToS is interpreted as DSCP and ECN  bits,  ECN  part
+          must be zero. It may also be the word inherit, in which case the
+          ToS will be copied from the inner packet if it is IPv4  or  IPv6
+          (otherwise  it  will be 0). The ECN fields are always inherited.
+          Default is 0.
+```
+
+which when set to inherit will take the marking from the inner header and apply it to the GENEVE packet. Therefore, east west traffic within the cluster can benefit
+from this API and not just the north south traffic.
+
+### API Details
+
+* A new API `NetworkQoS` under the `k8s.ovn.org/v1` version will be added to `go-controller/pkg/crd/networkqos/v1`.
+This would be a namespace-scoped CRD:
+
+```go
+import (
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=networkqoses
+// +kubebuilder::singular=networkqos
+// +kubebuilder:object:root=true
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=".status.status"
+// +kubebuilder:subresource:status
+// NetworkQoS is a CRD that allows the user to define a DSCP marking and metering
+// for pods ingress/egress traffic on its namespace to specified CIDRs,
+// protocol and port. Traffic belong these pods will be checked against
+// each Rule in the namespace's NetworkQoS, and if there is a match the traffic
+// is marked with relevant DSCP value and enforcing specified policing/shaping
+// parameters.
+type NetworkQoS struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   Spec   `json:"spec,omitempty"`
+	Status Status `json:"status,omitempty"`
+}
+
+// Spec defines the desired state of NetworkQoS
+type Spec struct {
+	// netAttachRefs points to a list of NetworkAttachmentDefinition object,
+	// to which the QoS Rules will be applied. The netAttachRefs can be of
+	// type Layer-3, Layer-2 or Localnet. If not specified, the default cluster-wide
+	//  network will be used.
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="netAttachRefs is immutable"
+	// +kubebuilder:validation:XValidation:rule="self.all(nad, nad.kind == 'NetworkAttachmentDefinition')", message="\"kind\" of netAttachRef must be \"NetworkAttachmentDefinition\""
+	NetworkAttachmentRefs []corev1.ObjectReference `json:"netAttachRefs,omitempty"`
```

Comment:
The comments in L96 to L99 needs to change with now this field means `networkAttachmentRefs`.

netAttachRefs points to a list of objects which could be either NAD, UDN, or Cluster UDN. In the case of NAD, the network type could be fo type Layer-3, Layer-2, or Localnet. If not specified, then the primary network of the selected Pods wil be chosen.